prefab/curl-ssl/app/src/main/cpp/java_interop.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <jni.h>
+
+#include <cstdlib>
+#include <string>
+
+#include "logging.h"
+
+namespace curlssl {
+namespace jni {
+
+template <typename JavaType>
+struct JavaClass;
+
+template <>
+struct JavaClass<jstring> {
+  static constexpr char name[] = "java/lang/String";
+};
+
+template <typename... ToTypes>
+struct Convert;
+
+template <>
+struct Convert<jstring> {
+  static jstring from(JNIEnv* env, const std::string& from) {
+    return env->NewStringUTF(from.c_str());
+  }
+};
+
+template <typename JavaSubType>
+struct Convert<jobjectArray, JavaSubType> {
+  template <typename SubType>
+  static jobjectArray from(JNIEnv* env, const std::vector<SubType>& from) {
+    jobjectArray array = env->NewObjectArray(
+        from.size(), env->FindClass(JavaClass<JavaSubType>::name), nullptr);
+    for (size_t i = 0; i < from.size(); ++i) {
+      env->SetObjectArrayElement(array, i, env->NewStringUTF(from[i].c_str()));
+    }
+    return array;
+  }
+};
+
+template <>
+struct Convert<std::string> {
+  static std::string from(JNIEnv* env, const jstring& from) {
+    typedef std::unique_ptr<const char[], std::function<void(const char*)>>
+        JniString;
+
+    JniString cstr(env->GetStringUTFChars(from, nullptr),
+                   [=](const char* p) { env->ReleaseStringUTFChars(from, p); });
+
+    if (cstr == nullptr) {
+      LOG_ERROR("%s: GetStringUTFChars failed", __func__);
+      std::abort();
+    }
+
+    const jsize len = env->GetStringUTFLength(from);
```

Comment:
i think this is actually more expensive than just letting std::string do the strlen(), so not worth the extra line.